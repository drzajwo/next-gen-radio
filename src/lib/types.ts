export interface GetRadioStationsResponse {
  data: RadioStationInfo[];
}

export interface RadioStationInfo {
  id: string;
  namedId?: string;
  description: string;
  name: string;
  imgUrl: string;
  streamUrl: string;
  reliability: number;
  popularity: number;
  tags: RadioStationTag[];
}

// TODO: it would be best to have known list of all possible tags
type RadioStationTag = string;
