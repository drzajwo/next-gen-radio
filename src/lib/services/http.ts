export const getData = async <T>(URL: string): Promise<T> => {
  return await (
    await fetch(URL, {
      method: "GET",
    })
  ).json();
};

// TODO: implement other HTTP methods when needed
