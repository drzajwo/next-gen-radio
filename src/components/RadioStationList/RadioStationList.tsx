import { Grid, GridCol } from "@mantine/core";

import { RadioStationCard } from "../RadioStationCard";

import classes from "./RadioStationList.module.scss";

import { getRadioStations } from "@/lib/services";

export const RadioStationList = async () => {
  const radioStationsList = await getRadioStations();

  return (
    <Grid
      gutter={{ base: 5, xs: "md", md: "xl", xl: 50 }}
      align="stretch"
      data-testid="radio-grid"
    >
      {radioStationsList.map((radioStationInfo, index) => (
        <GridCol
          span={{ base: 12, xs: 6, sm: 6, md: 4, lg: 2 }}
          key={`${radioStationInfo.id}_${index}`}
          className={classes.gridItem}
        >
          <RadioStationCard radioStationInfo={radioStationInfo} />
        </GridCol>
      ))}
    </Grid>
  );
};
