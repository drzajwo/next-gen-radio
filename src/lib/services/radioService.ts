import "server-only";

import type { GetRadioStationsResponse, RadioStationInfo } from "../types";

import { getData } from "./http";

const TUNE_IN_BASE_URL =
  "https://s3-us-west-1.amazonaws.com/cdn-web.tunein.com";

export const getRadioStations = async (): Promise<RadioStationInfo[]> => {
  try {
    const { data } = await getData<GetRadioStationsResponse>(
      `${TUNE_IN_BASE_URL}/stations.json`,
    );

    return data;
  } catch (e) {
    return [];
  }
};
