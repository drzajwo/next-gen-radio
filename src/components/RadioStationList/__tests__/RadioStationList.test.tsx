import { MantineProvider } from "@mantine/core";
import { render } from "@testing-library/react";

import "@testing-library/jest-dom";

import { RadioStationList } from "../RadioStationList";

import { getRadioStations } from "@/lib/services";

window.matchMedia = jest.fn();
jest.mock("@/lib/services", () => ({
  getRadioStations: jest.fn(),
}));
jest.mock("../../RadioStationCard", () => ({
  RadioStationCard: jest.fn().mockRejectedValue("RadioStationCard"),
}));
jest.mock("@mantine/core", () => ({
  ...jest.requireActual("@mantine/core"),
  GridCol: jest.fn(() => <div>Grid col</div>),
}));

const RenderUI = async () => (
  <MantineProvider>{await RadioStationList()}</MantineProvider>
);

describe("RadioStationList test", () => {
  it("renders a grid with items", async () => {
    (getRadioStations as jest.Mock).mockResolvedValue([
      {
        id: "s135217",
        namedId: "Radio Mozart_s135217",
        description: "Only Mozart Music",
        name: "Radio Mozart",
        imgUrl: "https://cdn-radiotime-logos.tunein.com/s135217d.png",
        streamUrl: "https://streamingv2.shoutcast.com/Radio-Mozart",
        reliability: 68,
        popularity: 2.9,
        tags: ["music", "classical"],
      },
    ]);

    const { getByTestId, container } = render(await RenderUI());

    expect(getByTestId("radio-grid")).toBeInTheDocument();
    expect(getRadioStations).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
