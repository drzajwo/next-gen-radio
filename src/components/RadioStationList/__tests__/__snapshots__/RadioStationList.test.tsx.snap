// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioStationList test renders a grid with items 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <style
    data-mantine-styles="inline"
  >
    .__m__-r0{--grid-gutter:calc(0.3125rem * var(--mantine-scale));}@media(min-width: 36em){.__m__-r0{--grid-gutter:var(--mantine-spacing-md);}}@media(min-width: 62em){.__m__-r0{--grid-gutter:var(--mantine-spacing-xl);}}@media(min-width: 88em){.__m__-r0{--grid-gutter:calc(3.125rem * var(--mantine-scale));}}
  </style>
  <div
    class="m_410352e9 mantine-Grid-root __m__-r0"
    data-testid="radio-grid"
    style="--grid-align: stretch;"
  >
    <div
      class="m_dee7bd2f mantine-Grid-inner"
    >
      <div>
        Grid col
      </div>
    </div>
  </div>
</div>
`;
